app/src/main/java/org/sil/gatherwords/SessionActivity.java

```diff
@@ -118,18 +128,110 @@ public void save_settings_fab_pressed(View view) {
         startActivity(i);
     }
 
-    private static class InsertSessionsTask extends AsyncTask<Session, Void, Void> {
+    private static class InsertSessionsTask extends AsyncTask<Session, Double, Void> {
         private SessionDao sDAO;
+        private WordDao wordDao;
+        private WeakReference<SessionActivity> sessionActivityReference;
+        private Long sessionID;
+        private int maxProgress;
+        private AppDatabase db;
+        private String wordList;
 
-        InsertSessionsTask(AppDatabase db) {
+        InsertSessionsTask(SessionActivity activity) {
+            sessionActivityReference = new WeakReference<>(activity);
+            db = AppDatabase.get(sessionActivityReference.get().getApplicationContext());
```

Comment:
No need for indirection here; we can just `db = AppDatabase.get(activity.getApplicationContext());`.